Source/GSAutoLayoutEngine.m

```diff
@@ -1,60 +1,460 @@
    version 2.1 of the License, or (at your option) any later version.
-   
+
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
-   
+
    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free
    Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110 USA.
 */
 
 #import "GSAutoLayoutEngine.h"
-#include "AppKit/NSLayoutConstraint.h"
+#import "AppKit/NSLayoutConstraint.h"
+#import "NSViewPrivate.h"
+#import "GSCSConstraint.h"
+#import "GSFastEnumeration.h"
+
+@interface
+GSAutoLayoutEngine (PrivateMethods)
+
+- (GSCSConstraint *) solverConstraintForConstraint:
+    (NSLayoutConstraint *)constraint;
+
+- (void) mapLayoutConstraint: (NSLayoutConstraint *)layoutConstraint
+          toSolverConstraint: (GSCSConstraint *)solverConstraint;
+
+- (void) updateAlignmentRectsForTrackedViews;
+
+- (void) updateAlignmentRectsForTrackedViewsForSolution:
+    (GSCSSolution *)solution;
+
+- (BOOL) hasConstraintsForView: (NSView *)view;
+
+- (GSCSConstraint *) getExistingConstraintForAutolayoutConstraint:
+    (NSLayoutConstraint *)constraint;
+
+- (void) addConstraintAgainstViewConstraintsArray:
+    (NSLayoutConstraint *)constraint;
+
+- (void) removeConstraintAgainstViewConstraintsArray:
+    (NSLayoutConstraint *)constraint;
+
+- (void) addSupportingInternalConstraintsToView: (NSView *)view
+                                   forAttribute: (NSLayoutAttribute)attribute
+                                     constraint: (GSCSConstraint *)constraint;
+
+- (void) removeInternalConstraintsForView: (NSView *)view;
+
+- (void) addSolverConstraint: (GSCSConstraint *)constraint;
+
+- (void) removeSolverConstraint: (GSCSConstraint *)constraint;
+
+- (NSArray *) constraintsForView: (NSView *)view;
+
+- (NSNumber *) indexForView: (NSView *)view;
+
+- (void) notifyViewsOfAlignmentRectChange: (NSArray *)viewsWithChanges;
+
+- (BOOL) solverCanSolveAlignmentRectForView: (NSView *)view
+                                   solution: (GSCSSolution *)solution;
+
+- (void) recordAlignmentRect: (NSRect)alignmentRect
+                forViewIndex: (NSNumber *)viewIndex;
+
+- (NSRect) solverAlignmentRectForView: (NSView *)view
+                             solution: (GSCSSolution *)solution;
+
+- (NSRect) currentAlignmentRectForViewAtIndex: (NSNumber *)viewIndex;
+
+- (BOOL) updateViewAligmentRect: (GSCSSolution *)solution view: (NSView *)view;
+
+@end
 
 @implementation GSAutoLayoutEngine
 
-- (void) addConstraint: (NSLayoutConstraint *)constraint
+- (instancetype) initWithSolver: (GSCassowarySolver *)cassowarySolver;
 {
-  // FIXME Add constraint to solver
+  self = [super init];
+  if (self != nil)
+    {
+      _solver = cassowarySolver;
```

Comment:
We really prefer to use ASSIGN here. In this specific case it will result in exactly the same code, but it better expresses the meaning of the operation.